docs: continuous deployment (#2019)
woodpecker-ci/woodpecker: docs/docs/20-usage/33-continuous-deployment.md

> # Continuous Deployment

Reviewer: In general I would like this is a nice overview, but I think its to generic to be actually helpful. I think we could rather add a new category: cookbook #316 or articles to our  new blog with concrete tutorial styled guides. So something like to deploy a docker container with ansible you would first build the container with WP, add it to a registry and then run an ansible script to deploy the new container.

A typical CI pipeline contains steps such as: *clone*, *build*, *test*, *package* and *push*. The final build product may be binaries pushed to a git repository or a docker container pushed to a container registry.

When these should be deployed on an app server, the pipeline should include a *deploy* step, which represents the "CD" in CI/CD - the automatic deployment of a pipeline's final product.

There are various ways to accomplish CD with Woodpecker, depending on your project's specific needs.

## Invoking deploy script via SSH

> The final step in your pipeline could SSH into the app server, and run a deployment script.

Reviewer 2: ```suggestion
The final step in your pipeline could SSH into the app server and run a deployment script.
```

One of the benefits would be that the deployment script's output could be included in the pipeline's log. However in general, this is a complicated option as it tightly couples the CI and app servers.

> An SSH step could be written as a Woodpecker plugin; although there is no official ssh plugin, one does exist for [Drone](https://plugins.drone.io/plugins/ssh) and could be adapted accordingly.

Reviewer 2: ```suggestion
An SSH step could be written by using an plugin, like [ssh](https://plugins.drone.io/plugins/ssh) or [git push](https://woodpecker-ci.org/plugins/Git%20Push).
```

## Polling for asset changes

This option completely decouples the CI and app servers, and there is no explicit deploy step in the pipeline.

> On the app server, one should create a script or cron job which polls for asset changes (every minute, say). When a new version is detected, the script redeploys the app.

Reviewer 2: ```suggestion
On the app server, one should create a script or cron job that polls for asset changes (every minute, say). When a new version is detected, the script redeploys the app.
```

This option is easy to maintain, but the downside is a short delay (one minute) before new assets are detected.

## Using a configuration management tool

> If you are using a configuration management tool (e.g. Ansible, Chef, Puppet), then you could setup the last pipeline step to call that tool so as to perform the redeployment.

Reviewer 2: ```suggestion
If you are using a configuration management tool (e.g. Ansible, Chef, Puppet), then you could setup the last pipeline step to call that tool to perform the redeployment.
```

> A [Drone](https://plugins.drone.io/plugins/ansible) plugin for Ansible exists and could be adapted accordingly.

Reviewer 2: ```suggestion
A plugin for [Ansible](https://plugins.drone.io/plugins/ansible) exists and could be adapted accordingly.
```

This option is complex and only suitable in an environment in which you're already using configuration management.

> ## Using webhooks (recommended)

Reviewer: Not sure what this has to do with WP and why we should recommend it?

> If your forge (Github, Gitlab, Gitea, etc.) supports webhooks, then you could create a separate listening app that receives a webhook when new assets are available, and redeploys your app.

Reviewer 2: ```suggestion
If your forge (Github, Gitlab, Gitea, etc.) supports webhooks, then you could create a separate listening app that receives a webhook when new assets are available and redeploys your app.
```

The listening "app" can be something as simple as a PHP script.

> Alternatively, there are a number of popular webhook servers which simplify this process, so you only need to write your actual deployment script. For example, [webhook](https://github.com/adnanh/webhook) and [webhookd](https://github.com/ncarlier/webhookd).

Reviewer 2: ```suggestion
Alternatively, there are a number of popular webhook servers that simplify this process, so you only need to write your actual deployment script. For example, [webhook](https://github.com/adnanh/webhook) and [webhookd](https://github.com/ncarlier/webhookd).
```
